src/Orleans.Runtime/Silo/Silo.cs

```diff
@@ -557,8 +531,57 @@ void StartGateway()
             if (logger.IsEnabled(LogLevel.Debug)) { logger.Debug("Silo.Start complete: System status = {0}", this.SystemStatus); }
         }
 
+        private async Task OnBecomeActiveStart(CancellationToken ct)
+        {
+            var stopWatch = Stopwatch.StartNew();
+            StartTaskWithPerfAnalysis("Start gateway", StartGateway, stopWatch);
+            void StartGateway()
+            {
+                // Now that we're active, we can start the gateway
+                var mc = this.messageCenter as MessageCenter;
+                mc?.StartGateway(this.Services.GetRequiredService<ClientObserverRegistrar>());
```

Comment:
We could stick this into ClientObserverRegistrar instead and make it responsible for adding itself to the MessageCenter by subscribing to `BecomeActive`.

The comment on `BecomeActive` suggests that only `MembershipOracle` should subscribe to it, but we're doing other things here.

Reply:
It's fine to leave this, though - we can untangle gateway initialization later. Maybe just update the `BecomeActive` comment to also include the gateway.

Reply:
For membership and gateway, we want to do it explicitly in Silo.cs to ensure that they are started in this order. I agree that I can add this in the comment